### PR TITLE
HIVE-22538: RS deduplication does not always enforce hive.optimize.reducededuplication.min.reducer

### PR DESCRIPTION
**Pull request by kasakrisz**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcRecordUpdater.java [hunk 1]
```diff
@@ -579,6 +579,7 @@ public void close(boolean abort) throws IOException {
 
             OrcFile.WriterOptions wo = OrcFile.writerOptions(this.options.getConfiguration())
                 .inspector(rowInspector)
+                .setSchema(createEventSchemaFromTableProperties(this.options.getTableProperties()))
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcRecordUpdater.java]: Why is this needed after this patch?

File: ql/src/test/results/clientpositive/acid_table_directories_test.q.out [hunk 2]
```diff
@@ -154,6 +154,7 @@ POSTHOOK: Input: default@acidparttbl@p=200
 ### ACID BASE DIR ###
 ### ACID BASE DIR ###
 ### ACID BASE DIR ###
+### ACID BASE DIR ###
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/acid_table_directories_test.q.out]: Is this expected?

File: ql/src/test/results/clientpositive/autoColumnStats_4.q.out [hunk 3]
```diff
@@ -128,16 +127,20 @@ STAGE PLANS:
             Statistics: Num rows: 10 Data size: 1728 Basic stats: COMPLETE Column stats: COMPLETE
             Group By Operator
               aggregations: compute_stats(a, 'hll'), compute_stats(b, 'hll')
-              minReductionHashAggr: 0.99
-              mode: hash
+              mode: complete
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/autoColumnStats_4.q.out]: Why is the GBy operator changing to complete mode here? That could potentially slow down the insert. Does it have anything to do with the size (numrows) of the insert?

File: ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out [hunk 4]
```diff
@@ -261,31 +260,21 @@ STAGE PLANS:
                   Statistics: Num rows: 1 Data size: 4 Basic stats: COMPLETE Column stats: NONE
                   Group By Operator
                     aggregations: compute_stats(a, 'hll'), compute_stats(b, 'hll')
-                    minReductionHashAggr: 0.99
-                    mode: hash
+                    mode: complete
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out]: Same as above... complete here would imply a single instance to write data to the table, which could slow down execution. What is the root cause for this?

File: ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out [hunk 5]
```diff
@@ -814,30 +791,20 @@ STAGE PLANS:
                   Statistics: Num rows: 1 Data size: 8 Basic stats: COMPLETE Column stats: NONE
                   Group By Operator
                     aggregations: compute_stats(a, 'hll'), compute_stats(b, 'hll')
-                    minReductionHashAggr: 0.99
-                    mode: hash
+                    mode: complete
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out]: It seems this complete mode is only happening for write type INSERT... I am wondering whether we are setting up the number of reducers to 1 too aggressively since INSERT does not need a single instance.

File: ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcRecordUpdater.java (hunk 1 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcRecordUpdater.java]: I simplified this part of the code: reuse the existing writerOptions instance which contains the correct schema.

File: ql/src/test/results/clientpositive/acid_table_directories_test.q.out (hunk 2 above)
Comment (kasakrisz) [reply ql/src/test/results/clientpositive/acid_table_directories_test.q.out]: Enabling parallelism when inserting eliminates this change.

File: ql/src/test/results/clientpositive/autoColumnStats_4.q.out (hunk 3 above)
Comment (kasakrisz) [reply ql/src/test/results/clientpositive/autoColumnStats_4.q.out]: Enabling parallelism when inserting eliminates this change.

File: ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out (hunk 4 above)
Comment (kasakrisz) [reply ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out]: Enabling parallelism when inserting eliminates this change.

File: ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out (hunk 5 above)
Comment (kasakrisz) [reply ql/src/test/results/clientpositive/llap/dynamic_semijoin_reduction_3.q.out]: Enabling parallelism when inserting eliminates this change.
